docs/proposals/004-scalability-benchmarking.md

```diff
@@ -0,0 +1,102 @@
+---
+title: Argo CD Scalability Benchmarking
+authors:
+  - "@andklee"
+  - "@csantanapr"
+  - "@morey-tech"
+  - "@nimakaviani"
+sponsors:
+  - 
+reviewers:
+  - 
+  - 
+approvers:
+  - 
+  - 
+creation-date: 2023-02-28
+last-updated: 2023-02-28
+---
+
+# Argo CD Scalability Benchmarking
+
+## Open Questions [optional]
+* What are the scalability factors to prioritize when benchmarking Argo CD?
+    * Reconciliation time for all, one, or a number of applications in the cluster.
+
+## Summary
+
+## Motivation
+Users of Argo CD are interested to know how to scale Argo CD, what configuration tweaks and deployment options they have, and how far they can push resources (in terms of the number of supported applications, Git repositories, managing clusters, etc.).
+
+While the Argo CD documentation [discusses options](https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#scaling-up) to scale up, the actual process is not clear and, as articulated [in this thread](https://github.com/argoproj/argo-cd/issues/9633), oftentimes a point of confusion for users.
+
+By running large-scale benchmarking, we aim at helping the Argo CD community with the following:
+
+* Give confidence to organizations running at significant scale (_needs to be quantified_) that Argo CD can support their use-case, with empirical evidence.
+* Create clear guidelines for scaling Argo CD based on the key scalability factors.
+* Provide recommendations for which topology is best suited for users based on their needs.
+* Determine what work, if any, is needed to improve the scalability of Argo CD.
+
+### Goals
+1. Create a standard set of repeatable benchmarking procedures to objectively measure the limitations of Argo CD.
+    1. May result in a new `argo-cd-benchmarking` repo under `argoproj-labs` so that anyone can easily replicate it and the development of the procedures can happen outside of the lifecycle for Argo CD (unlike [the current `gen-resources` hack](https://github.com/argoproj/argo-cd/tree/master/hack/gen-resources) in the main project).
+    2. Include detailed test scenarios that account for key scalability factors that allow for easy tweaking of the parameters to simplify testing of alternative scenarios.
+2. Determine the baseline for when tweaking is required on the default configuration (resource allocations and replicas).
+    1. One cluster, Applications in In-cluster, default resource allocations.
+3. Quantify how tweaking of existing parameters (replicas, sharding, parallelism, etc) impacts the performance of Argo CD.
+4. Provide a set of metrics and thresholds that will provide a basis for automatically scaling the Argo CD components and alerting for when performance is being impacted by limitations.
+
```

Comment:
5. All recommendations, examples and scenarios will be vendor agnostic.

Reply:
```suggestion
5. All tooling, recommendations, examples, and scenarios will be vendor-agnostic.
```